dynacred/src/builder.ts

```diff
@@ -0,0 +1,274 @@
+                    (id) => this.retrieveDarcBS(id),
+                ))),
+                map((dbs) => dbs.filter((db) => db !== undefined)),
+                map((dbs) => dbs.map((db) => new ActionBS(db))),
+            )));
+        return new AddressBook(contactBS, groupsBS, actionsBS);
+    }
+
+    async retrieveCoinBS(coinID: BehaviorSubject<InstanceID> | InstanceID):
+        Promise<CoinBS> {
+        Log.lvl3("getting coinBS");
+        if (coinID instanceof Buffer) {
+            coinID = new BehaviorSubject(coinID);
+        }
+        const coinObs = coinID.pipe(
+            flatMap((id) => this.bc.instanceObservable(id)),
+            mergeAll(),
+            map((inst) => CoinInstance.create(this.bc as any, inst.key, inst.darcID, Coin.decode(inst.value))),
+        );
+        return new CoinBS(await ObservableToBS(coinObs));
+    }
+
+    async retrieveCredentialSignerBS(credStructBS: CredentialStructBS): Promise<CredentialSignerBS> {
+        const signerDarcID = (await this.retrieveSignerIdentityDarc(credStructBS.darcID)).id;
+        const darcBS = await this.retrieveDarcBS(signerDarcID);
+        Log.lvl3("getting devices");
+        const devices = await this.retrieveCST(darcBS, credStructBS.credDevices, "device");
+        Log.lvl3("getting recoveries");
+        const recoveries = await this.retrieveCST(darcBS, credStructBS.credRecoveries, "recovery");
+        return new CredentialSignerBS(darcBS, devices, recoveries);
+    }
+
+    async retrieveCST(signerBS: DarcBS, cimbs: CredentialInstanceMapBS, prefix: string): Promise<CSTypesBS> {
+        Log.lvl3("getting CS:", prefix);
+        const aisbs = ConvertBS(cimbs, (im) => im.toInstanceIDs());
+        return new CSTypesBS(signerBS, cimbs, prefix, await this.retrieveDarcsBS(aisbs));
+    }
+
+    async retrieveCredentialStructBS(id: InstanceID): Promise<CredentialStructBS | undefined> {
+        Log.lvl3("creating CredentialStruct from scratch:", id);
+        try {
+            const instBS = await this.bc.instanceObservable(id);
+            const darcID = instBS.getValue().darcID;
+            const credBS = ConvertBS(instBS, (inst) => CredentialStruct.decode(inst.value));
+            return new CredentialStructBS(id, darcID, credBS);
+        } catch (e) {
+            Log.warn("couldn't get credStruct for", id, e);
+            return undefined;
+        }
+    }
+
+    async retrieveDarcsBS(aisbs: BehaviorSubject<InstanceID[]>): Promise<DarcsBS> {
+        Log.lvl3("getting darcsBS");
+        const darcs = await ObservableToBS(aisbs.pipe(
+            flatMap((ais) => Promise.all(ais
+                .map((iid) => this.retrieveDarcBS(iid)))),
+            map((dbs) => dbs.filter((db) => db !== undefined)),
+        ));
+        return new DarcsBS(darcs);
+    }
+
+    async retrieveDarcBS(darcID: BehaviorSubject<InstanceID> | InstanceID):
+        Promise<DarcBS | undefined> {
+        Log.lvl3("getting darcBS");
+        if (darcID instanceof Buffer) {
```

Comment:
yes? Why not? if it's a `BehaviorSubject`, then it can change over time (which is probably rare, in fact). If it's an `InstanceID` (which is a `Buffer`), then it's fixed.

Reply:
> `InstanceID` (which is a `Buffer`)

Aha, that was the part I was missing.
Shouldn't it then be `if ... instanceof InstanceID` for clarity?

Reply:
Furthermore (but this is personal preference), instead of calling a function with various possible different types and hoping the function does the right thing, I find it clearer to either have a polymorphic type, or do:
```ts
async retrieveDarcBSwithID(darcID: InstanceID): Promise<DarcBS | undefined> {
    return await retrieveDarcBS(new BehaviorSubject(darcID));
}

async retrieveDarcBS(darcID: BehaviorSubject<InstanceID>): Promise<DarcBS | undefined> {
    ...
}
```

Reply:
Yes, using two methods is the go way. But I like polymorphism ;)

I tried `darcID instanceof InstanceID`, but it doesn't work. So I added a comment:
```
        // Need to verify against Buffer here, which is the defined type of InstanceID.
        // Else typescript complains....
```

Reply:
> Yes, using two methods is the go way. But I like polymorphism ;)

But in an OO language, the polymorphism usually employs a class hierarchy, where a different subtype implements a common method in their appropriate way: this removes the need to explicitly check for the type in the using code, and is transparently extended if new subtypes appear.

> I tried `darcID instanceof InstanceID`, but it doesn't work.

That's another bad point for TypeScript ;-)  Just out of curiosity, what does the error say?

> So I added a comment:
> 
> ```
>         // Need to verify against Buffer here, which is the defined type of InstanceID.
>         // Else typescript complains....
> ```

Thank you, better than nothing. :)

Reply:
```
TS2693: 'InstanceID' only refers to a type, but is being used as a value here. 
```

And, yes, that's not polymorphism.